datafusion/core/src/physical_optimizer/enforce_distribution.rs

```diff
@@ -927,9 +927,8 @@ fn add_hash_on_top(
     n_target: usize,
     repartition_beneficial_stats: bool,
 ) -> Result<DistributionContext> {
-    let partition_count = input.plan.output_partitioning().partition_count();
     // Early return if hash repartition is unnecessary
-    if n_target == partition_count && n_target == 1 {
```

Comment:
I think previous check was wrong, partition count equality doesn't mean their hash requirement is satisfied. However, when target is 1 hash requirement is trivially satisfied